CODING_GUIDELINES.md

```diff
@@ -1,37 +1,186 @@
+- All of your SCSS styles should be wrapped in an `@include exports('XXX.clarity') { ... }`, to avoid producing duplicates in the deliverable CSS.
+- The baseline should always be respected to ensure proper vertical rhythm.
+
+
+## Typescript and Angular
+
+### Public API
+
+- Custom texts should almost always be received through content projection, **not** inputs. This will allow for icons, images, links, ...
+- Offer an output for every internal state change, whether it's triggered by a user action or an internal operation (expand/collapse, on/off, option selected, ...).
+- Offer two-way binding rather than simple outputs when possible.
+- Use structural directives to receive `TemplateRef` instances, in order to offer a less verbose API than `<ng-template>` to the consumer.
+- Have related components (like parent / child) communicate by default through services.
+- Never use explicitly named template reference variables as part of the API, to query them through `@ContentChild`. Simply receive structural directives or inputs.
+- Never make HTTP calls, and never impose formatting constraints on their backend API.
+
+Finally, and this is harder to put in black-or-white terms, keep the API as simple and natural as possible, even if it means more work on Clarity's side. Implementing multiple directives and components and making them communicate through services will lead to a much more pleasant integration than forcing your consumer to pass data manually from one to the other. Here are a couple examples of what to do and not to do (taken from real-life carousel components):
+- **Don't:** Using an `<ng-template>` and passing an `$implicit` value to it, that the consumer will have to tie to a template local variable:
+```html
+<clr-carousel [clrItems]="items">
+  <ng-template let-item>
+    <h1>{{item}}</h1>
+  </ng-template>
+</clr-carousel>
+```
+- **Don't:** Receive an input on the "wrong" component, and force the user to handle the communication between this component and the other, linked ones (in this case through a template reference variable):
+```html
+<clr-carousel #carousel [clrItems]="items">
+  <h1 *ngFor="let item of carousel.visible">
+    {{item}}
+  </h1>
+</clr-carousel>
+```
+- **Do:** Introduce a new structural directive that behaves like an `*ngFor`, but will automatically communicate with the parent component through services:
+```html
+<clr-carousel>
+  <h1 *clrVisible="let item of items">
+    {{item}}
+  </h1>
+</clr-carousel>
+```
+
+
+### Code style
+
+- If the template is just a few lines, it should go inline in the component's metadata with `template`. Otherwise, it should be in a separate HTML file using `templateUrl`.
+- Bindings that are always true should be declared directly in the component's metadata using the `host` option. Other bindings should be on the corresponding property or getter, using the `@HostBinding` annotation.
```

Comment:
https://angular.io/guide/styleguide#hostlistenerhostbinding-decorators-versus-host-metadata this goes against what angular recommends, no?

Reply:
Yes, we're doing our own thing on that. adding a class to the host with Angular is a non-pattern for applications, but we have to do it all the time as a library. That's why we have the exception of the "always true" bindings.